edge/pkg/edged/edged_status.go

```diff
@@ -436,13 +445,122 @@ func (e *edged) updateNodeStatus() error {
 		return err
 	}
 
+	now := time.Now()
+	if features.DefaultMutableFeatureGate.Enabled(features.NodeLease) {
+		// If enable NodeLease feature gate, check if it's the time to update nodestatus or
+		// if the nodestatus has changed since last report. If it's not in any situation of both,
+		// skip update nodestaus.
+		if now.Before(e.lastStatusReportTime.Add(e.nodeStatusUpdateFrequency)) && !e.nodeStatusHasChanged(&nodeStatus.Status) {
+			return nil
+		}
+	}
 	err = e.metaClient.NodeStatus(e.namespace).Update(e.nodeName, *nodeStatus)
 	if err != nil {
 		klog.Errorf("update node failed, error: %v", err)
+		return err
 	}
 	e.setInitNode(&v1.Node{
 		Status: *nodeStatus.Status.DeepCopy(),
 	})
+	e.lastReportStatus = nodeStatus.Status.DeepCopy()
+	e.lastStatusReportTime = time.Now()
+	return nil
+}
+
+func (e *edged) nodeStatusHasChanged(currentStatus *v1.NodeStatus) bool {
+	originalStatusCopy := e.lastReportStatus.DeepCopy()
+	currentStatusCopy := currentStatus.DeepCopy()
+
+	// compare NodeStatus.Condition
+	if len(originalStatusCopy.Conditions) != len(currentStatusCopy.Conditions) {
+		return true
+	}
+	sort.SliceStable(originalStatusCopy.Conditions, func(i, j int) bool {
+		return originalStatusCopy.Conditions[i].Type < originalStatusCopy.Conditions[j].Type
+	})
+	sort.SliceStable(currentStatusCopy.Conditions, func(i, j int) bool {
+		return currentStatusCopy.Conditions[i].Type < currentStatusCopy.Conditions[j].Type
+	})
+	emptyHeartbeatTime := metav1.Time{}
+	for i := range currentStatusCopy.Conditions {
+		// ignore difference of LastHeartbeatTime
+		originalStatusCopy.Conditions[i].LastHeartbeatTime = emptyHeartbeatTime
+		currentStatusCopy.Conditions[i].LastHeartbeatTime = emptyHeartbeatTime
+		if !apiequality.Semantic.DeepEqual(originalStatusCopy.Conditions[i], currentStatusCopy.Conditions[i]) {
+			return true
+		}
+	}
+
+	// compare other fields of NodeStatus
+	originalStatusCopy.Conditions, currentStatusCopy.Conditions = nil, nil
+	return !apiequality.Semantic.DeepEqual(originalStatusCopy, currentStatusCopy)
+}
+
+func (e *edged) newLease(base *coordinationv1.Lease) *coordinationv1.Lease {
```

Comment:
in edge,  we need implement interface

```
	Create(ctx context.Context, lease *v1.Lease, opts metav1.CreateOptions) (*v1.Lease, error)
	Update(ctx context.Context, lease *v1.Lease, opts metav1.UpdateOptions) (*v1.Lease, error)
	Get(ctx context.Context, name string, opts metav1.GetOptions) (*v1.Lease, error)
```

Reply:
Well, I considered of just using `k8s.io/component-helpers/apimachinery/lease`. I decided not to use it beacuse it depends deeply at the apierrors returned back from API Server. For example:
https://github.com/kubernetes/component-helpers/blob/master/apimachinery/lease/controller.go#L189-L192
https://github.com/kubernetes/component-helpers/blob/master/apimachinery/lease/controller.go#L155-L172

It uses optimistic lock machanism of API Server. It always try to update lease first. If there're some conflicts, it then get the new lease and update again. Its call stack is:
tryUpdateLease----Conflict---->Get node lease----NotFound---->create node lease

So if we use `k8s.io/component-helpers/apimachinery/lease`, we have to implement the similar apierrors message. Otherwise we have to modify the machanism of this package. Neither is easy.

Reply:
ok, got it, if so, I suggest putting the nodeLease related code in a single file

Reply:
agree.